system_test/helpers/aest_docker.erl

```diff
@@ -0,0 +1,260 @@
+-export([start/1]).
+-export([stop/1]).
+-export([setup_node/2]).
+-export([delete_node/1]).
+-export([start_node/1]).
+-export([stop_node/1, stop_node/2]).
+-export([kill_node/1]).
+-export([get_service_address/2]).
+
+%=== MACROS ====================================================================
+
+-define(CONFIG_FILE_TEMPLATE, "epoch.yaml.mustache").
+-define(EPOCH_CONFIG_FILE, "/home/epoch/epoch.yaml").
+-define(EPOCH_LOG_FOLDER, "/home/epoch/node/log").
+-define(EPOCH_MINE_RATE, 1000).
+-define(EXT_HTTP_PORT, 3013).
+-define(INT_HTTP_PORT, 3113).
+-define(INT_WS_PORT, 3114).
+
+%=== TYPES =====================================================================
+
+-type log_fun() :: fun((io:format(), list()) -> ok) | undefined.
+-type test_uid() :: binary() | undefined.
+-type service_label() :: ext_http | int_http | int_ws.
+
+%% State of the docker backend
+-type backend_state() :: #{
+    postfix := binary(),        % A unique postfix to add to container names.
+    log_fun := log_fun(),       % Function to use for logging.
+    data_dir := binary(),       % The directory where the templates can be found.
+    temp_dir := binary(),       % A temporary directory that can be used to generate
+                                % configuration files and save the log files.
+    net_id := binary()          % Docker network identifier
+}.
+
+%% Node specification
+-type node_spec() :: #{
+    name := atom(),
+    % Names or URLs of the peer nodes
+    peers := [atom() | binary()],
+    source := {pull, binary()}  % Source of the node image
+}.
+
+
+%% State of a node
+-type node_state() :: #{
+    spec := map(),              % Backup of the spec used when adding the node
+    log_fun := log_fun(),       % Function to use for logging
+    hostname := atom(),         % Hostname of the container running the node
+    exposed_ports := #{service_label() => pos_integer()},
+    local_ports := #{service_label() => pos_integer()}
+}.
+
+
+%=== GENERIC API FUNCTIONS =====================================================
+
+-spec start(Options) -> State
+    when Options :: #{test_id => test_uid(),
+                      log_fun => log_fun(),
+                      data_dir := binary(),
+                      temp_dir := binary()},
+         State :: backend_state().
+
+start(Options) ->
+    TestId = maps:get(test_id, Options),
+    Postfix = uid2postfix(TestId),
+    LogFun = maps:get(log_fun, Options),
+    {ok, DataDir} = maps:find(data_dir, Options),
+    {ok, TempDir} = maps:find(temp_dir, Options),
+    ok = aest_docker_api:start(),
+    NetName = <<"epoch", Postfix/binary>>,
+    #{'Id' := NetId} = aest_docker_api:create_network(#{name => NetName}),
+    log(LogFun, "Network ~p [~s] created", [NetName, NetId]),
+    #{postfix => Postfix,
+      log_fun => LogFun,
+      data_dir => DataDir,
+      temp_dir => TempDir,
+      net_id => NetId
+    }.
+
+
+-spec stop(BackendState) -> ok
+    when BackendState :: backend_state().
+
+stop(BackendState) ->
```

Comment:
I would prefer a more compact spec declaration format:

```
-spec stop(backend_state()) -> ok.
stop(BackendState) ->
    ...
```

It has less duplication (of variable names) and doesn't feel like a separate function declaration.

Reply:
For small functions like this one it looks a bit more verbose, but for complicated specs it looks way cleaner that way (at least for me), and I prefer uniformity... In addition, I think that having the variable names usually helps (for more complex functions)...

Reply:
For me it's just one more point where code rot can start to creep in, where you forget to change one name but not the other. Making it more compact would save 2-4 lines per function. Just my 2 cents 🙂

Reply:
It would save lines only for small functions, for bigger one it makes them cleaner (IMHO). About code rote, dialyzer would not detect variable names being wrong, but it would detect function typing getting wrong forcing you to change the spec...

Reply:
In this case even `-spec stop(BackendState :: backend_state() -> ok.` fits in one line. Breaking it over several lines is _not_ helping.

And *no*, unfortunately, dialyzer won't find squat 90% of the time in this situation... Dialyzer will just use the success type!?

Reply:
@sylane That was exactly my point of code rot, I'm arguing for removing the variable names. Keeping the variable name in two places with no compiler check in-between is not good. It's enough just to use the type spec.

Reply:
@hanssv I usually go by the rule that if `typer` doesn't find anything to complain about, I just stick with the deduced types.